// microsoft-adal-angular6/esm5/src/ms-adal-angular6.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { MsAdalAngular6Service } from './ms-adal-angular6.service';
var MsAdalAngular6Module = /** @class */ (function () {
    function MsAdalAngular6Module() {
    }
    /**
     * @param {?} adalConfig
     * @return {?}
     */
    MsAdalAngular6Module.forRoot = /**
     * @param {?} adalConfig
     * @return {?}
     */
    function (adalConfig) {
        return {
            ngModule: MsAdalAngular6Module,
            providers: [MsAdalAngular6Service, { provide: 'adalConfig', useValue: adalConfig }]
        };
    };
    MsAdalAngular6Module.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    declarations: [],
                    exports: []
                },] },
    ];
    return MsAdalAngular6Module;
}());
export { MsAdalAngular6Module };
function MsAdalAngular6Module_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    MsAdalAngular6Module.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    MsAdalAngular6Module.ctorParameters;
}
